ethexe/contracts/src/Router.sol

```diff
@@ -26,515 +22,417 @@ contract Router is IRouter, OwnableUpgradeable, ReentrancyGuardTransient {
 
-        require(router.codes[codeId] == CodeState.Unknown, "code with such id already requested or validated");
+        require(
+            router.protocolData.codes[_codeId] == Gear.CodeState.Unknown,
+            "given code id is already on validation or validated"
+        );
 
-        router.codes[codeId] = CodeState.ValidationRequested;
+        router.protocolData.codes[_codeId] = Gear.CodeState.ValidationRequested;
 
-        emit CodeValidationRequested(codeId, blobTxHash);
+        emit CodeValidationRequested(_codeId, _blobTxHash);
     }
 
-    function createProgram(bytes32 codeId, bytes32 salt, bytes calldata payload, uint128 _value)
+    function createProgram(bytes32 _codeId, bytes32 _salt, bytes calldata _payload, uint128 _value)
         external
-        payable
         returns (address)
     {
-        (address actorId, uint128 executableBalance) = _createProgramWithoutMessage(codeId, salt, _value);
+        (address actorId, uint128 executableBalance) = _createProgramWithoutMessage(_codeId, _salt, _value);
 
-        IMirror(actorId).initMessage(tx.origin, payload, _value, executableBalance);
+        IMirror(actorId).initMessage(msg.sender, _payload, _value, executableBalance);
 
         return actorId;
     }
 
     function createProgramWithDecoder(
-        address decoderImplementation,
-        bytes32 codeId,
-        bytes32 salt,
-        bytes calldata payload,
+        address _decoderImpl,
+        bytes32 _codeId,
+        bytes32 _salt,
+        bytes calldata _payload,
         uint128 _value
-    ) external payable returns (address) {
-        (address actorId, uint128 executableBalance) = _createProgramWithoutMessage(codeId, salt, _value);
+    ) external returns (address) {
+        (address actorId, uint128 executableBalance) = _createProgramWithoutMessage(_codeId, _salt, _value);
 
         IMirror mirrorInstance = IMirror(actorId);
 
-        mirrorInstance.createDecoder(decoderImplementation, keccak256(abi.encodePacked(codeId, salt)));
+        mirrorInstance.createDecoder(_decoderImpl, keccak256(abi.encodePacked(_codeId, _salt)));
 
-        mirrorInstance.initMessage(tx.origin, payload, _value, executableBalance);
+        mirrorInstance.initMessage(msg.sender, _payload, _value, executableBalance);
 
         return actorId;
     }
 
-    function commitCodes(CodeCommitment[] calldata codeCommitmentsArray, bytes[] calldata signatures) external {
-        Storage storage router = _getStorage();
-
-        bytes memory codeCommetmentsHashes;
-
-        for (uint256 i = 0; i < codeCommitmentsArray.length; i++) {
-            CodeCommitment calldata codeCommitment = codeCommitmentsArray[i];
+    // # Validators calls.
+    function commitCodes(Gear.CodeCommitment[] calldata _codeCommitments, bytes[] calldata _signatures) external {
+        Storage storage router = _router();
+        require(router.genesisBlock.hash != bytes32(0), "router genesis is zero; call `lookupGenesisHash()` first");
```

Comment:
Should be moved to a separate modifier "fullyInitialized()"

Reply:
Impossible to do keeping no-op. modifier can access arguments for free, but here we need to have `Storage storage router` query of it consumes gas, so repeating code seems a bit more optimal